include/swift/ABI/TaskLocal.h

```diff
@@ -101,20 +109,55 @@ class TaskLocal {
     /// the Item linked list into the appropriate parent.
     static Item *createParentLink(AsyncTask *task, AsyncTask *parent);
 
+    static Item *createLink(AsyncTask *task,
+                            const HeapObject *key,
+                            const Metadata *valueType,
+                            bool inTaskGroupBody);
+
     static Item *createLink(AsyncTask *task,
                             const HeapObject *key,
                             const Metadata *valueType);
 
+    static Item *createLinkInTaskGroup(
+        AsyncTask *task,
+        const HeapObject *key,
+        const Metadata *valueType);
+
     void destroy(AsyncTask *task);
 
     Item *getNext() {
       return reinterpret_cast<Item *>(next & ~statusMask);
     }
 
+    void relinkNext(Item* nextOverride) {
+      assert(!getNext() &&
+               "Can only relink task local item that was not pointing at anything yet");
+      assert(nextOverride->isNextLinkPointer() ||
+             nextOverride->isParentPointer() &&
+                 "Currently relinking is only done within a task group to "
+                 "avoid within-taskgroup next pointers; attempted to point at "
+                 "task local declared within task group body though!");
```

Comment:
`A || B && C` is parsed as `A || (B && C)`.  That turns out to be logically equivalent in this case, but still, it's better to get into the habit of parenthesizing assertion conditions.

The assertion and behavior here is strange — the method name suggests that it's just changing the `next` pointer, but you're really assuming a stronger condition.  It feels like you can't really assert that condition properly, which to me is usually a sign that the operation isn't as general as I'm thinking it is.  That's okay; operations don't always generalize.  It just suggests that you ought to name this method something that indicates its special behavior, and maybe feel less bad about writing something that doesn't feel like a generic operation.

I assume what we're doing here is cloning the entries and then calling `relinkNext` on them.

Reply:
I think it's a good point to rename this method, it is a very specific trick. When we detect the "next" task local item is "safe" as in "not the bad pattern where we have to copy" we immediately link to it rather than continuing our copying dance. I'll rename and see if I can clarify some more.

Fixing the assertion too, thank you for noticing that

Reply:
cleaned up in https://github.com/apple/swift/pull/74089